src/thunderbots/software/radio_communication/mrf/dongle.cpp

```diff
@@ -0,0 +1,610 @@
+            }
+        }
+        pan_ = DEFAULT_CONFIGS[config].pan;
+        {
+            const char *pan_string = std::getenv("MRF_PAN");
+            if (pan_string)
+            {
+                int i = std::stoi(pan_string, nullptr, 0);
+                if (i < 0 || i > 0xFFFE)
+                {
+                    throw std::out_of_range(
+                        "PAN must be between 0x0000 (0) and 0xFFFE (65,534).");
+                }
+                pan_ = static_cast<uint16_t>(i);
+            }
+        }
+        device.control_no_data(LIBUSB_REQUEST_TYPE_VENDOR | LIBUSB_RECIPIENT_INTERFACE,
+                               MRF::CONTROL_REQUEST_SET_CHANNEL, channel_,
+                               static_cast<uint16_t>(radio_interface), 0);
+        device.control_no_data(LIBUSB_REQUEST_TYPE_VENDOR | LIBUSB_RECIPIENT_INTERFACE,
+                               MRF::CONTROL_REQUEST_SET_SYMBOL_RATE,
+                               symbol_rate == 625 ? 1 : 0,
+                               static_cast<uint16_t>(radio_interface), 0);
+        device.control_no_data(LIBUSB_REQUEST_TYPE_VENDOR | LIBUSB_RECIPIENT_INTERFACE,
+                               MRF::CONTROL_REQUEST_SET_PAN_ID, pan_,
+                               static_cast<uint16_t>(radio_interface), 0);
+        device.control_out(LIBUSB_REQUEST_TYPE_VENDOR | LIBUSB_RECIPIENT_INTERFACE,
+                           MRF::CONTROL_REQUEST_SET_MAC_ADDRESS, 0,
+                           static_cast<uint16_t>(radio_interface), &MAC, sizeof(MAC), 0);
+
+        {
+            std::chrono::system_clock::time_point now = std::chrono::system_clock::now();
+            std::chrono::system_clock::time_point epoch =
+                std::chrono::system_clock::from_time_t(0);
+            std::chrono::system_clock::duration diff = now - epoch;
+            std::chrono::microseconds micros =
+                std::chrono::duration_cast<std::chrono::microseconds>(diff);
+            uint64_t stamp = static_cast<uint64_t>(micros.count());
+            device.control_out(LIBUSB_REQUEST_TYPE_VENDOR | LIBUSB_RECIPIENT_DEVICE,
+                               MRF::CONTROL_REQUEST_SET_TIME, 0, 0, &stamp, sizeof(stamp),
+                               0);
+        }
+    }
+
+    // Switch to normal mode.
+    device.set_interface_alt_setting(radio_interface, normal_altsetting);
+
+    // Prepare the available message IDs for allocation.
+    for (unsigned int i = 0; i < 256; ++i)
+    {
+        free_message_ids.push(static_cast<uint8_t>(i));
+    }
+
+    // Submit the message delivery report transfers.
+    for (auto &i : mdr_transfers)
+    {
+        i.reset(new USB::BulkInTransfer(device, 1, 8, false, 0));
+        i->signal_done.connect(sigc::mem_fun(this, &MRFDongle::handle_mdrs));
+        i->submit();
+    }
+
+    // Submit the received message transfers.
+    for (auto &i : message_transfers)
+    {
+        i.reset(new USB::BulkInTransfer(device, 2, 105, false, 0));
```

Comment:
What do these numbers mean?

Reply:
Seems like 2 is the endpoint number, 105 is the maximum expected length of bytes to receive, 0 is timeout  (just going by the BulkInTransfer header)

Reply:
Could we add a quick comment to this effect?